lib/Sema/PreCheckExpr.cpp

```diff
@@ -1510,6 +1537,58 @@ TypeExpr *PreCheckExpression::simplifyUnresolvedSpecializeExpr(
   return nullptr;
 }
 
+// Checks whether this is a BinaryExpr with operator `&` and returns the
+// BinaryExpr, if so.
+static BinaryExpr *getCompositionExpr(Expr *expr) {
+  if (auto *binaryExpr = dyn_cast<BinaryExpr>(expr)) {
+    // look at the name of the operator, if it is a '&' we can create the
+    // composition TypeExpr
+    auto fn = binaryExpr->getFn();
+    if (auto Overload = dyn_cast<OverloadedDeclRefExpr>(fn)) {
+      for (auto Decl : Overload->getDecls())
+        if (Decl->getBaseName() == "&") {
+          return binaryExpr;
+        }
+    } else if (auto *Decl = dyn_cast<UnresolvedDeclRefExpr>(fn)) {
+      if (Decl->getName().isSimpleName() &&
+          Decl->getName().getBaseName() == "&")
+        return binaryExpr;
+    }
+  }
+
+  return nullptr;
+}
+
+bool PreCheckExpression::isValidPlaceholderTypeExprParent(Expr *E) {
+  if (CallArgs.count(E) > 0) return false;
+
+  // A TupleExpr can only be a valid parent if all the elements are potentially
+  // valid TypeExprs, or if it's the single element of a dictionary expr.
+  if (auto *tupleExpr = dyn_cast<TupleExpr>(E)) {
+    if (llvm::all_of(tupleExpr->getElements(), [&](Expr *elt) {
+      return isa<TypeExpr>(elt) || isa<DiscardAssignmentExpr>(elt);
+    }))
+      return true;
+
+    if (ExprStack.size() >= 2)
+      if (auto *DE = dyn_cast<DictionaryExpr>(ExprStack[ExprStack.size() - 2]))
+        if (DE->getElements().size() == 1)
+          return true;
+  }
+
+  if (auto *arrayExpr = dyn_cast<ArrayExpr>(E))
+    if (arrayExpr->getElements().size() == 1)
+      return true;
+
+  return E->isValidTypeExprParent() ||
+      isa<OptionalEvaluationExpr>(E) ||
+      isa<BindOptionalExpr>(E) ||
+      isa<ForceValueExpr>(E) ||
+      isa<ParenExpr>(E) ||
+      isa<ArrowExpr>(E) ||
+      getCompositionExpr(E);
+}
```

Comment:
This hack helps us preserve some diagnostics for uses of '`_`' which were probably "supposed" to be patterns/discard assignments, but it's ugly. Would love any suggestions for a better way to handle this!

Reply:
Hmm, we could maybe switch from the bottom-up type folding we're currently doing to a top-down approach. 🤔

Reply:
Actually, I think there's a better heuristic we can use here that involves keeping track of information the whole way down the tree.